xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/js/xwiki/table/livetable.js

```diff
@@ -303,7 +304,10 @@ XWiki.widgets.LiveTable = Class.create({
 
     for (var i = off; i <= f; i++) {
       if (this.fetchedRows[i]) {
-        var elem = this.handler(this.fetchedRows[i], i, this);
+        if (this.renderedRows[i] === undefined) {
+          this.renderedRows[i] = this.handler(this.fetchedRows[i], i, this);
+        }
+        var elem = this.renderedRows[i];
```

Comment:
The idea of caching the DOM rendering produced by the `handler()` method.

However, I think it's not good to cache the replies in a different array that is not synthconized with the data `fetchedRows`. If you follow `fetchedRows`, you will see that it gets updated in various situtations and, if `renderedRows` is not updated as well, each time, you will most likely end up showing stale renderings in the UI that no longer correspond to the new data inside `fetchedRows`.

What I propose instead is that you set some special `this.fetchedRows[i].livetableRenderedRow` property (hoping no LivetableResults implementation actually defines a property named `livetableRenderedRow` in its JSON results rows) where you can store the rendering of individual `fetchedRows`. This way, whenever a data `fetchedRow` is removed or added, its rendering will stay close to it and be added/removed as well.

The only downside I can see in this proposed approach would be that if some `xwiki:livetable:newrow` event handlers or even the row renderer callback would want to modify the cached data row that it receives as parameter, its modification would not be reflected in the UI, unless maybe if it explicitly clears the cached rendering (by clearing `row.livetableRenderedRow` and maybe recalling `displayRows`). Anyway, I'm not sure this is supported right now either, so, IMO, it's safe to assume that the rendering done by the `this.handler` (i.e. javascript callback) method is cacheable. If you want to be extra safe, you could add a livetable option that controls if rendering caching is performed or not (enabled by default), so as to not block anyone that might be doing weird client side hacks. I personally don't think it's worth the complication, since we don't support that by design.

Reply:
Was also thinking about javascript event listeners that might be registered on the row renderings that get cached. You'd need to check and make sure that the listeners are still attached and working once you start adding and removing rendered rows multiple times on the DOM.

Reply:
AFAIK javascript event listeners bind to an element are not automatically removed by the gc until the element itself is removed. In our case, removing a row from the DOM doesn't make the gc remove it as long as it is still referenced in our cache.
I'll make sure this is the case.